src/Microsoft.Kusto.ServiceLayer/DataSource/KustoClient.cs

```diff
@@ -221,11 +231,20 @@ public IDataReader ExecuteQuery(string query, CancellationToken cancellationToke
         /// </summary>
         /// <param name="query">The query.</param>
         /// <returns>The results.</returns>
-        public Task<IDataReader> ExecuteQueryAsync(string query, CancellationToken cancellationToken,
+        public async Task<IEnumerable<T>> ExecuteQueryAsync<T>(string query, CancellationToken cancellationToken,
             string databaseName = null)
         {
-            var reader = ExecuteQuery(query, cancellationToken, databaseName);
-            return Task.FromResult(reader);
+            try
+            {
+                var resultReader = await Task.Run(() => ExecuteQuery(query, cancellationToken, databaseName), cancellationToken);
+                var results = KustoDataReaderParser.ParseV1(resultReader, null);
+                var tableReader = results[WellKnownDataSet.PrimaryResult].Single().TableData.CreateDataReader();
+                return new ObjectReader<T>(tableReader);
+            }
+            catch (Exception)
+            {
+                return null;
```

Comment:
It will be good to add telemetry on these failures so that we can trace what went wrong.